components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/IdentityProviderManager.java

```diff
@@ -1724,6 +1757,69 @@ public void addIdP(IdentityProvider identityProvider, String tenantDomain)
         }
     }
 
+    /**
+     * Adds an Identity Provider to the given tenant
+     *
+     * @param identityProvider new Identity Provider information
+     * @throws IdentityProviderManagementException Error when adding Identity Provider
+     *                                             information
+     */
+    @Override
+    public IdentityProvider addIdPWithResourceId(IdentityProvider identityProvider, String tenantDomain)
+            throws IdentityProviderManagementException {
+
+        // invoking the pre listeners
+        Collection<IdentityProviderMgtListener> listeners = IdPManagementServiceComponent.getIdpMgtListeners();
+        for (IdentityProviderMgtListener listener : listeners) {
+            if (listener.isEnable() && !listener.doPreAddIdP(identityProvider, tenantDomain)) {
+                return null;
+            }
+        }
+
+        int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
+
+        if (IdPManagementServiceComponent.getFileBasedIdPs().containsKey(identityProvider.getIdentityProviderName())
+                && !identityProvider.getIdentityProviderName().startsWith(IdPManagementConstants.SHARED_IDP_PREFIX)) {
+            //If an IDP with name starting with "SHARED_" is added from UI, It's blocked at the service class
+            // before calling this method
+            throw IdPManagementUtil.handleClientException(IdPManagementConstants.ErrorMessage
+                    .ERROR_CODE_IDP_ALREADY_EXISTS, identityProvider.getIdentityProviderName());
+        }
+
+        if (identityProvider.getPermissionAndRoleConfig() != null
+                && identityProvider.getPermissionAndRoleConfig().getRoleMappings() != null) {
+            verifyAndUpdateRoleConfiguration(tenantDomain, tenantId, identityProvider.getPermissionAndRoleConfig());
+        }
+
+        if (IdentityProviderManager.getInstance().getIdPByName(
+                identityProvider.getIdentityProviderName(), tenantDomain, true) != null) {
+            throw IdPManagementUtil.handleClientException(IdPManagementConstants.ErrorMessage
+                    .ERROR_CODE_IDP_ALREADY_EXISTS, identityProvider.getIdentityProviderName());
+        }
+        String idpName = identityProvider.getIdentityProviderName();
+        String metadata = handleMetadta(identityProvider);
+
+        validateIdPEntityId(identityProvider.getFederatedAuthenticatorConfigs(), tenantId, tenantDomain);
+        if (idpName.length() > 0 && metadata.length() > 0) {
+            if (SAML2SSOMetadataConverter != null) {
```

Comment:
change the SAML2SSOMetadataConverter variable to camel case.

Reply:
created a separate issue [1] to track metadata improvements. This issue will be fixed in it.
[1] wso2/product-is#6750